### DISCUSSION TITLE
middleware api is hard to hook when need to through request response

### DISCUSSION THREAD

**Discussion by Germiniku** [go-resty/resty#578] [Ideas]
The scenario is that I need to build tracking into the middleware, but it's hard to do this without intruding through BeforeRequest and AfterRequest hooks
 If middleware api is like this, I can do a lot of things with closures
```
Middleware func(*Client, *Request) func(*Client, *Response) error
```

**Comment by kecci** [comment]
Hi @Germiniku,

What I do on tracking with resty, I should set my resty client with this func:
- Hook BeforeRequest, use`restyClient.OnAfterResponse()`
- Hook AfterRequest, use `restyClient.OnBeforeRequest()`

I hope this could be help the discussion.
